### DISCUSSION TITLE
Validate characters of filter expression on SetFilter

### DISCUSSION THREAD

**Discussion by Arthurvdv** [StefanMaron/BusinessCentral.LinterCop#464] [New Rule]
A colleague of my wrote a `SetFilter` and expected to use the wildcard sign in the filter expression itself.

Bad code
```AL
Vendor.SetFilter("VAT Registration No.", '*%1*', EnterpriseVATNo);
```

Good code
```AL
Vendor.SetFilter("VAT Registration No.", '%1', '*' + EnterpriseVATNo + '*');
```

> The filter expression. A valid expression consists of alphanumeric characters and one or more of the following operators: <, >, \, &, |, and =. You can use replacement fields (%1, %2, and so on) to insert values at run-time.

[Record.SetFilter(Any, Text [, Any,...]) Method - Business Central | Microsoft Learn](https://learn.microsoft.com/en-us/dynamics365/business-central/dev-itpro/developer/methods-auto/record/record-setfilter-method)

**Comment by rvanbekkum** [comment]
Why do it that way though?

**Reply by Arthurvdv** [reply]
I'm not familiar with a better way of handling wildcards within a `SetFilter`.

```AL
Contact.SetFilter(Name, '%1%2%3%2', '@', '*', MyVar);
```

The examples I find are like this, is there a better way of handling this?

**Reply by rvanbekkum** [reply]
If you want to check that the VAT Registration No. contains a certain value, then your first example is ok as well, right?

```al
Vendor.SetFilter("VAT Registration No.", '*%1*', EnterpriseVATNo);
```

Maybe it becomes a different story if you would like to check for the presence of reserved/special characters like '*'.

**Reply by Arthurvdv** [reply]
I believe that's not the case, I've tested this with the example below

```AL
pageextension 50100 "Customer List" extends "Customer List"
{
    trigger OnOpenPage()
    var
        Customer: Record Customer;
        PartialSearchValue: Text;
    begin
        // Customer 10000 (The Cannon Group PLC) in CRONUS International Ltd. has a "VAT Registration No." with the value 789456278
        PartialSearchValue := '8945627';

        Customer.SetFilter("VAT Registration No.", '*%1*', PartialSearchValue);
        Message('A: Found %1 customers', Customer.Count());

        Customer.SetFilter("VAT Registration No.", '%1', '*' + PartialSearchValue + '*');
        Message('B: Found %1 customers', Customer.Count());
    end;
}
```

The results of this is example is
```
A: Found 0 customers
B: Found 1 customers
```

My conclusion was that the wildcard shouldn't be in the filter expression itself. Or am I missing something here? 🤔

**Reply by jwikman** [reply]
I believe you are right @Arthurvdv 

The way this usually is implemented (at least a few times in BaseApp) is with `StrSubstNo()`
```al
        Customer.SetFilter("VAT Registration No.", StrSubstNo('*%1*', PartialSearchValue));
```
Which is, IMO, more elegant than concatenating strings. 🙂

**Reply by rvanbekkum** [reply]
Aah I see, thanks for explaining further. 🙂

**Reply by jwikman** [reply]
This code should also cover other special characters for filters, as `@` and `?`. 

Bad code
```
Customer.SetFilter("VAT Registration No.", '*%1*', PartialSearchValue);
Customer.SetFilter("VAT Registration No.", '@*%1*', PartialSearchValue);
Customer.SetFilter("VAT Registration No.", '??%1*', PartialSearchValue);
```

Good code
```
Customer.SetFilter("VAT Registration No.", StrSubstNo('*%1*', PartialSearchValue));
Customer.SetFilter("VAT Registration No.", StrSubstNo('@*%1*', PartialSearchValue));
Customer.SetFilter("VAT Registration No.", StrSubstNo('??%1*', PartialSearchValue));
```

Good code (a bit unnecessary parameter, but it works)
```
Customer.SetFilter("VAT Registration No.", '%1', StrSubstNo('*%1*', PartialSearchValue));
Customer.SetFilter("VAT Registration No.", '%1', StrSubstNo('@*%1*', PartialSearchValue));
Customer.SetFilter("VAT Registration No.", '%1', StrSubstNo('??%1*', PartialSearchValue));
```

The rule would probably only need to check if any of `*`, `?` or `@` is found in the second parameter of `SetFilter()` together with any placeholder (`%1`, `%2` etc), don't you think? If you then decide to use string concatenation or `StrSubstNo()` is up to the developer, it's more a matter of taste. 🙂
But in the rule description you might want to mention `StrSubstNo()`, as that one works as both 2nd and 3rd parameter and should be easier to remember.

**Reply by Arthurvdv** [reply]
Thank you for your insights!

For now I didn't include a validation if a placeholder (%1, %2 etc) is there. In the example below it could be that you want to filter on a specific length of a field. Then I also want the rule to be raised.
```AL
Customer.SetFilter("VAT Registration No.", '???');
```

The new [LC0050](https://github.com/StefanMaron/BusinessCentral.LinterCop/wiki/LC0050) _SetFilter with unsupported operator in filter expression_ rule is now included in the (pre)release version of [v0.30.11](https://github.com/StefanMaron/BusinessCentral.LinterCop/releases/tag/v0.30.11) of the LinterCop.

P.S. Your examples where perfect for the documentation and I've taken the liberty to use them in the wiki :-)

**Reply by jwikman** [reply]
@Arthurvdv I believe that you need to include the placeholders in the rule, I think that they are the key to where the SetFilter issue comes from.

This is perfectly fine to use:

```
Customer.SetFilter("VAT Registration No.", '???');
Customer.SetFilter("VAT Registration No.", 'SE*');
```
While this is not
```
Customer.SetFilter("VAT Registration No.", '%1?', '??');
Customer.SetFilter("VAT Registration No.", '%1*', 'SE');
```
Or have you seen some other behavior that I've missed?

**Reply by Arthurvdv** [reply]
```AL
pageextension 50100 "Customer List" extends "Customer List"
{
    trigger OnOpenPage()
    var
        Customer: Record Customer;
    begin
        Customer.SetFilter("VAT Registration No.", '*8');
        Message('A: Found %1 customers', Customer.Count());

        Customer.SetFilter("VAT Registration No.", '*%1', '8');
        Message('B: Found %1 customers', Customer.Count());
    end;
}
```

```
A: Found 4 customers
B: Found 0 customers
```

🤯 

You're right @jwikman, I was fully convinced that A would also return zero.
Is it really that the operator switches from operator to literal when a %1 is included in the string?

**Reply by jwikman** [reply]
That is the behavior I can see... But it is indeed my own conclusions after testing, I have not seen any documentation on this.

**Reply by Arthurvdv** [reply]
https://github.com/StefanMaron/BusinessCentral.LinterCop/pull/492/files

I've added the check for any placeholder (%1, %2 etc) now to the rule (including a typo threaded vs treated 😳)

Thank you for pointing this out, I wasn't aware of this behavior with an without an placeholder.

**Reply by jwikman** [reply]
😅

**Reply by jwikman** [reply]
And immediately this rule helped me to fix a bug that has been in my code for a year or two! 👍

**Reply by Arthurvdv** [reply]
That's awesome!

**Comment by Arthurvdv** [comment]
[Behavior of selected operators on SetFilter together with a placeholder (%1) · Issue #7631 · microsoft/AL (github.com)](https://github.com/microsoft/AL/issues/7631)

I've created an issue for an follow-up of this to determine if this behavior is an bug or a design choice.
